compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -636,53 +639,66 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
+        self.assemble_probe_inner(generalized_self_ty);
+    }
 
+    fn assemble_probe_inner(&mut self, generalized_self_ty: Ty<'tcx>) {
+        debug!("assemble_probe: self_ty={:?}", generalized_self_ty);
+        match *generalized_self_ty.kind() {
+            ty::Dynamic(data, ..) if let Some(p) = data.principal() => {
                 self.assemble_inherent_candidates_from_object(generalized_self_ty);
                 self.assemble_inherent_impl_candidates_for_type(p.def_id());
                 if self.tcx.has_attr(p.def_id(), sym::rustc_has_incoherent_inherent_impls) {
-                    self.assemble_inherent_candidates_for_incoherent_ty(raw_self_ty);
+                    self.assemble_inherent_candidates_for_incoherent_ty(generalized_self_ty);
                 }
             }
             ty::Adt(def, _) => {
                 let def_id = def.did();
                 self.assemble_inherent_impl_candidates_for_type(def_id);
                 if self.tcx.has_attr(def_id, sym::rustc_has_incoherent_inherent_impls) {
-                    self.assemble_inherent_candidates_for_incoherent_ty(raw_self_ty);
+                    self.assemble_inherent_candidates_for_incoherent_ty(generalized_self_ty);
                 }
             }
             ty::Foreign(did) => {
                 self.assemble_inherent_impl_candidates_for_type(did);
                 if self.tcx.has_attr(did, sym::rustc_has_incoherent_inherent_impls) {
-                    self.assemble_inherent_candidates_for_incoherent_ty(raw_self_ty);
+                    self.assemble_inherent_candidates_for_incoherent_ty(generalized_self_ty);
                 }
             }
             ty::Param(p) => {
                 self.assemble_inherent_candidates_from_param(p);
             }
+            ty::Alias(ty::Projection, projection_ty) if self.tcx.trait_solver_next() => {
+                let new_infer_ty = self.next_ty_var(TypeVariableOrigin {
+                    kind: TypeVariableOriginKind::NormalizeProjectionType,
+                    span: self.span,
+                });
+                let obligation = traits::Obligation::new(
+                    self.tcx,
+                    self.misc(self.span),
+                    self.param_env,
+                    ty::Binder::dummy(ty::ProjectionPredicate {
+                        projection_ty,
+                        term: new_infer_ty.into(),
+                    }),
+                );
+                if self.predicate_may_hold(&obligation) {
+                    let mut fulfill_cx = <dyn TraitEngine<'tcx>>::new_in_snapshot(self.tcx);
+                    fulfill_cx.register_predicate_obligation(self, obligation);
+                    let errors = fulfill_cx.select_where_possible(self);
+                    if errors.is_empty() {
+                        self.assemble_probe_inner(self.resolve_vars_if_possible(new_infer_ty))
+                    }
+                }
```

Comment:
i think here, and in `structurally_resolve_type`, and somewhere else :grin: we want to reuse this code. Can you move it to a function in the `InferCtxt` which you call before the match?